vault/identity_store_entities.go

```diff
@@ -475,7 +490,79 @@ func (i *IdentityStore) pathEntityNameDelete() framework.OperationFunc {
 	}
 }
 
-func (i *IdentityStore) handleEntityDeleteCommon(ctx context.Context, txn *memdb.Txn, entity *identity.Entity) error {
+// pathEntityIDDelete deletes the entity for a given entity ID
+func (i *IdentityStore) handleEntityBatchDelete() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+		entityIDs := d.Get("entity_ids").([]string)
+		if len(entityIDs) == 0 {
+			return logical.ErrorResponse("missing entity ids to delete"), nil
+		}
+
+		// Sort the ids by the bucket they will be deleted from
+		byBucket := make(map[string]map[string]struct{})
+		for _, id := range entityIDs {
+			bucket, ok := byBucket[i.entityPacker.BucketKey(id)]
+			if !ok {
+				bucket = make(map[string]struct{})
+			}
+
+			bucket[id] = struct{}{}
+			byBucket[i.entityPacker.BucketKey(id)] = bucket
+		}
+
+		deleteIdsForBucket := func(entityIDs []string) error {
+			i.lock.Lock()
```

Comment:
Since we are locking the entire identity store here. I am not sure if we are buying any gains here.

Reply:
What I meant to say was that we might be able to process a few operations between updates to different buckets. But when the batch deletes are trying to acquire this lock in quick succession, I am not sure if we'll be able to get a lot of things done between the lock switches.

Reply:
I think we should continue to release the lock between bucket writes otherwise we would lock all of the identity store for a long period of time (writing N buckets)

Reply:
I agree with Brian, Vault needs to be able to continue to serve login requests during these long operations or it will cause goroutines to pile up and kill the server. We've witnessed that a few times with the existing merge entities API.